### PR TITLE
[Feat][SM90] Add cutlass int4 x bf16 MoE

### PR DESCRIPTION
**Pull request by IwakuraRein**
## Purpose

- Add cutlass int4 x bf16 MoE by referencing #29691 .
- Added an environmental variable `VLLM_WNA16_MOE_BACKEND: Literal["cutlass", "marlin", "triton"] = "marlin"`

## Test Plan

**Unit Test**

- `pytest tests/kernels/quantization/test_cutlass_w4a16_moe.py`
- `pytest tests/kernels/quantization/test_cutlass_w4a16_moe_mm.py`

**E2E Test**

- `VLLM_WNA16_MOE_BACKEND=cutlass vllm serve RedHatAI/Qwen3-30B-A3B-quantized.w4a16 --trust-remote-code --max-num-seqs 1024 --max-num-batched-tokens 8192`
  `python tests/evals/gsm8k/gsm8k_eval.py`

**Performance Test**

- vllm bench serve --model RedHatAI/Qwen3-Next-80B-A3B-Instruct-quantized.w4a16 --trust-remote-code --ignore-eos --num-prompts 4096 --max-concurrency 4096 --random-input-len 8192 --random-output-len 1024 --random-range-ratio 0

## Test Result

**Unit Test**

- Unit tests passed

**E2E Accuracy Test**

```
Results:
Accuracy: 0.892
Invalid responses: 0.000
Total latency: 28.004 s
Questions per second: 47.100
Total output tokens: 150304
Output tokens per second: 5367.170
```

**Performance Test**

Cutlass path shows better perf when ISL is large.

`Qwen3-30B-A3B-quantized.w4a16`, 4k/1k, H200

**Marlin**

```
Output token throughput (tok/s):         3274.78
Peak output token throughput (tok/s):    5617.00
Total token throughput (tok/s):          16373.88
```

**Cutlass**

```
Output token throughput (tok/s):         3598.40
Peak output token throughput (tok/s):    5817.00
Total token throughput (tok/s):          17992.01
```

`Qwen3-Next-80B-A3B-Instruct-quantized.w4a16`, 8k/1k, H200 x 2 (TP)

**Marlin**

```
Output token throughput (tok/s):         2823.25
Peak output token throughput (tok/s):    11357.00
Total token throughput (tok/s):          25409.21
```

**Cutlass**

```
Output token throughput (tok/s):         3000.24
Peak output token throughput (tok/s):    32465.00
Total token throughput (tok/s):          27002.16
```

**Kernel Benchmark**

Tested on a single H200

| hidden size | intermediate size | expert num | topk | batch size | Marlin | Cutlass | delta |
|-------------|-------------------|------------|------|------------|--------|---------|-------|
| 3072        | 3072              | 64         | 4    | 1          | 50.77  | 53.73   | 94%   |
| 3072        | 3072              | 64         | 4    | 32         | 114.77 | 132.578 | 87%   |
| 3072        | 3072              | 64         | 4    | 128        | 211.11 | 309.588 | 68%   |
| 3072        | 3072              | 64         | 4    | 256        | 323.83 | 498.493 | 65%   |
| 3072        | 3072              | 64         | 4    | 512        | 527.16 | 457.407 | 115%  |
| 3072        | 3072              | 128        | 4    | 1          | 51.3   | 54.83   | 94%   |
| 3072        | 3072              | 128        | 4    | 32         | 115.7  | 121.1   | 96%   |
| 3072        | 3072              | 128        | 4    | 128        | 339.9  | 282.66  | 120%  |
| 3072        | 3072              | 128        | 4    | 256        | 383.9  | 526.85  | 73%   |
| 3072        | 3072              | 128        | 4    | 512        | 579.2  | 959     | 60%   |
| 7168        | 2048              | 48         | 8    | 1          | 48.1   | 63.1    | 76%   |
| 7168        | 2048              | 48         | 8    | 32         | 195.4  | 231.48  | 84%   |
| 7168        | 2048              | 48         | 8    | 128        | 315    | 479.41  | 66%   |
| 7168        | 2048              | 48         | 8    | 256        | 473.08 | 387.56  | 122%  |
| 7168        | 2048              | 96         | 8    | 1          | 47.7   | 63.48   | 75%   |
| 7168        | 2048              | 96         | 8    | 32         | 172.48 | 195.36  | 88%   |
| 7168        | 2048              | 96         | 8    | 128        | 485.17 | 409.59  | 118%  |

## TODO

- Update CUTLASS to allow flexible scale vector size. Then, relax the requirements for group size in both int4 x bf16 and int4 x fp8 MoE.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: csrc/quantization/cutlass_w4a8/get_group_starts.cuh
Comment (pavanimajety) [inline csrc/quantization/cutlass_w4a8/get_group_starts.cuh]:
```suggestion
// ElementGroupScale is cutlass::Array<cutlass::float_e4m3_t, 8> (packed fp8)
// or optional for W4A16
```

File: csrc/quantization/cutlass_w4a8/get_group_starts.cuh [hunk 1]
```diff
@@ -17,68 +17,111 @@ __global__ void get_group_gemm_starts(
...
   constexpr int pack_factor = 8;  // pack 8 int4 into int32
   b_offsets[expert_id] = b_base_as_int + (expert_id * k * n / pack_factor);
   b_group_scales_offsets[expert_id] =
-      b_group_scales_base_as_int + (expert_id * scale_k * n);
+      b_group_scales_base + (expert_id * scale_k * n);
 }
 
-#define __CALL_GET_STARTS_KERNEL(TENSOR_C_TYPE, C_TYPE)                  \
-  else if (out_tensors.dtype() == TENSOR_C_TYPE) {                       \
+#define __CALL_GET_STARTS_KERNEL_FP8(TENSOR_C_TYPE, C_TYPE)              \
```
Comment (pavanimajety) [inline csrc/quantization/cutlass_w4a8/get_group_starts.cuh:46]: Slightly confused here: Why do we call FP8?

File: vllm/_custom_ops.py [hunk 2]
```diff
@@ -1501,19 +1502,91 @@ def cutlass_w4a8_moe_mm(
...
+    )
+
+
 def cutlass_encode_and_reorder_int4b_grouped(
     b_tensors: torch.Tensor,
 ) -> tuple[torch.Tensor, torch.Tensor]:
     return torch.ops._C.cutlass_encode_and_reorder_int4b_grouped(b_tensors)
 
 
+def cutlass_reorder_int4b_grouped(
```
Comment (pavanimajety) [inline vllm/_custom_ops.py:1664]: Nit: Please add doc string for this

File: vllm/_custom_ops.py [hunk 3]
```diff
@@ -1501,19 +1502,91 @@ def cutlass_w4a8_moe_mm(
...
+    Executes the CUTLASS-based fused-MoE grouped matrix multiplication for the
+    W4A16 quantization scheme. Uses group-wise quantization (INT4 -> BFLOAT16)
+    and both per-channel + per-token scaling in the epilogue.
+
+    Args:
+        out_tensors:
+            Output buffer for all experts (updated in-place).
+        a_tensors:
+            BF16 activations for all experts.
+        b_tensors:
```
Comment (pavanimajety) [inline vllm/_custom_ops.py:1616]: Confused why the a_tensors here are Bfloat16 but the kernel calls them as Float8_e4m3?

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 4]
```diff
@@ -1178,3 +1178,397 @@ def cutlass_moe_w4a8_fp8(
...
+    """
+    This function computes a w4a16-quantized Mixture of Experts (MoE) layer
+    using two sets of quantized weights, w1_q and w2_q, and top-k gating
+    mechanism. The matrix multiplications are implemented with CUTLASS
+    mixed-dtype grouped gemm.
+
+    Parameters:
+    - a (torch.Tensor): The input tensor to the MoE layer.
+        Shape: [M, K]
+    - w1_q (torch.Tensor): The first set of bfloat16 expert weights.
```
Comment (pavanimajety) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py]: INT4 weights and fix the expected shape too?

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 5]
```diff
@@ -1178,3 +1178,397 @@ def cutlass_moe_w4a8_fp8(
...
+) -> torch.Tensor:
+    """
+    This function computes a w4a16-quantized Mixture of Experts (MoE) layer
+    using two sets of quantized weights, w1_q and w2_q, and top-k gating
+    mechanism. The matrix multiplications are implemented with CUTLASS
+    mixed-dtype grouped gemm.
+
+    Parameters:
+    - a (torch.Tensor): The input tensor to the MoE layer.
+        Shape: [M, K]
```
Comment (pavanimajety) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:1528]: Please add dtype for `a` as well

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 6]
```diff
@@ -1178,3 +1178,397 @@ def cutlass_moe_w4a8_fp8(
...
+    - w1_q (torch.Tensor): The first set of bfloat16 expert weights.
+        Shape: [num_experts, 2*N, K]
+    - w2_q (torch.Tensor): The second set of bfloat16 expert weights.
```
Comment (pavanimajety) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py]: same as above

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 7]
```diff
@@ -1178,3 +1178,397 @@ def cutlass_moe_w4a8_fp8(
...
+    assert w1.dtype == torch.int32
+    assert w2.dtype == torch.int32
+    assert w1.size(0) == w2.size(0), "Weights expert number mismatch"
+    assert out_dtype in [torch.bfloat16], f"Invalid output dtype: {out_dtype}"
+    if expert_map is not None:
+        assert expert_num_tokens is None
+    assert not use_batched_format, "batched format not supported yet"
+    assert group_size % 64 == 0, (
+        f"Only multiple of 64 is supported but got {group_size=}"
+    )
```
Comment (pavanimajety) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:1260]: Is this designed to not work for Kimi-K2? Any way we can support GS=32 for supporting Kimi-K2 as well?

**Review by pavanimajety — COMMENTED** [review]
Thanks for the PR @IwakuraRein, left a few early comments incase the PR is still WIP.

File: csrc/quantization/cutlass_w4a8/get_group_starts.cuh (hunk 1 above)
Comment (IwakuraRein) [reply csrc/quantization/cutlass_w4a8/get_group_starts.cuh:46]: It's exclusive to FP8 mma and FP8x8 scale

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 7 above)
Comment (IwakuraRein) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:1260]: There is an issue in the cutlass code base blocking GS==32. I will fix that in the near future.

File: vllm/_custom_ops.py (hunk 3 above)
Comment (IwakuraRein) [reply vllm/_custom_ops.py:1616]: Could you share the pointer to where a_tensors is marked as `Float8_e4m3`? I couldn't find it in `w4a16_grouped_mm_entry.cu`

Comment (pavanimajety) [reply vllm/_custom_ops.py:1616]: here in the dispatch: https://github.com/vllm-project/vllm/pull/33380/changes#diff-b2e359adc68749e45148d3b484cfb942ade47e8270bd1a9284fd1a4be8bd1007R47 

I think you are modifying the exisitng interface to take in both FP8 and BF16 a_tensors and dispatch to its corresponding case?

Comment (pavanimajety) [reply vllm/_custom_ops.py:1616]: If this is true, could we also run a small W4A8 model to ensure we aren't breaking anything?

Comment (IwakuraRein) [reply vllm/_custom_ops.py:1616]: If the activation is bf16, it will dispatch `__CALL_GET_STARTS_KERNEL`; if the activation is fp8, it will dispatch `__CALL_GET_STARTS_KERNEL_FP8`.

Comment (IwakuraRein) [reply vllm/_custom_ops.py:1616]: Was trying to test [czhu-cohere/Qwen3-30B-A3B-quantized.w4a8](https://huggingface.co/czhu-cohere/Qwen3-30B-A3B-quantized.w4a8) but encounter loading some parameters : (

Tested [Qwen3-30B-A3B-quantized.w4a16](https://huggingface.co/RedHatAI/Qwen3-30B-A3B-quantized.w4a16) and forced using w4a8 instead. The accuracy seems normal:

```
Results:
Accuracy: 0.873
Invalid responses: 0.000
Total latency: 27.660 s
Questions per second: 47.686
Total output tokens: 153365
Output tokens per second: 5544.654
```

File: tests/kernels/quantization/test_cutlass_w4a16_moe.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+from vllm.scalar_type import scalar_types
+from vllm.v1.worker.workspace import init_workspace_manager
+
+torch.random.manual_seed(42)
+device = torch.device("cuda:0")
+init_workspace_manager(device=device)
+
+
+@pytest.mark.skipif(
+    not current_platform.is_cuda(),
```
Comment (mgoin) [inline tests/kernels/quantization/test_cutlass_w4a16_moe.py]: also requires `current_platform.is_device_capability(90)`

File: tests/kernels/quantization/test_cutlass_w4a16_moe_mm.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+import pytest
+import torch
+from test_cutlass_w4a8_moe import cutlass_quantize
+
+from vllm import _custom_ops as ops
+from vllm.platforms import current_platform
+from vllm.scalar_type import scalar_types
+
+
+@pytest.mark.skipif(current_platform.is_rocm(), reason="Skip for rocm")
```
Comment (mgoin) [inline tests/kernels/quantization/test_cutlass_w4a16_moe_mm.py]: ditto

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 10]
```diff
@@ -178,28 +180,54 @@ def get_moe_method(
...
+                )
+
+            # otherwise, select WNA16 backend according to VLLM_WNA16_MOE_BACKEND
+            backend_ = envs.VLLM_WNA16_MOE_BACKEND
+            if backend_ == "cutlass":  # cutlass has most strict limitations
+                if (
+                    weight_quant.num_bits == 4
+                    and group_size % 64 == 0
+                    and current_platform.is_cuda()
+                    and current_platform.get_device_capability()[0] == 9
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: use `is_device_capability(90)`

**Review by mgoin — COMMENTED** [review]
Looks reasonable to me, assuming tuning will come along later. We really need to refactor the wna16 moe quant configs to be easier to iterate with kernels 😅 
Just nits with device capability

**Review by LucasWilkinson — COMMENTED** [review]
Thanks for the contribution, this is awesome! 

I think whenever adding new kernels we should strive to provide benchmarking results and turn it on by default in atleast one situation (can be HW or model specific); i.e. demonstrate a win over existing kernels in the introductory PR (or a stacked PR). The hope of enforcing this being that we can improve default performance of vLLM and reduce tribal knowledge of when to use which kernels (we unfortunately have a long history of excessive envvars/flags and kernels being added only to be used by a very small set of people aware of their existence).

**Review by LucasWilkinson — COMMENTED** [review]
ignore (double post due to GH having issues)

**Comment by xinli-sw** [comment]
I would also vote for improving Marlin perf instead - perhaps we could take this as a reference implementation for that?
